packages/tasit-action/src/TasitAct.test.js

```diff
@@ -0,0 +1,137 @@
+    // TasitAcount.create()
+    // > Acount { wallet: ..., metaTxInfos..., etc }
+    wallet = Account.createFromPrivateKey(
+      "0x11d943d7649fbdeb146dc57bd9cfc80b086bfab2330c7b25651dbaf382392f60"
+    );
+
+    simpleStorage = new Contract(contractAddress, contractABI);
+    expect(simpleStorage).to.exist;
+    expect(simpleStorage.address).to.equal(contractAddress);
+    expect(simpleStorage.getValue).to.exist;
+    expect(simpleStorage.setValue).to.exist;
+    // Events are not implemented yet
+    //expect(simpleStorage.ValueChanged).to.exist;
+  });
+
+  it("should throw error when instantiated with invalid args", async () => {
+    try {
+      new Contract();
+      assert(false, "constructor without address and ABI");
+
+      new Contract(contractAddress);
+      assert(false, "constructor without ABI");
+
+      new Contract("invalid address");
+      assert(false, "constructor with invalid address");
+
+      new Contract("invalid address", contractABI);
+      assert(false, "constructor with invalid address and valid ABI");
+
+      new Contract(contractAddress, "invalid abi");
+      assert(false, "constructor with valid address and invalid ABI");
+    } catch (e) {
+      assert(true);
+    }
+  });
+
+  it("should call a read-only contract method", async () => {
+    const value = await simpleStorage.getValue();
+    expect(value).to.exist;
+  });
+
+  it("should throw error when setting invalid wallet", async () => {
+    try {
+      simpleStorage = simpleStorage.setWallet();
+      assert(false, "setting no wallet");
+
+      simpleStorage = simpleStorage.setWallet("invalid wallet");
+      assert(false, "setting invalid wallet");
+    } catch (e) {
+      assert(true);
+    }
+  });
+
+  it("should throw error when calling write method without account/wallet/signer", async () => {
+    try {
+      simpleStorage.setValue("hello world");
+      assert(false, "calling write function without account");
+    } catch (e) {
+      assert(true);
+    }
+  });
+
+  it("should throw when subscribing with invalid trigger", async () => {
+    simpleStorage = simpleStorage.setWallet(wallet);
+    const subscription = simpleStorage.setValue("hello world");
+
+    try {
+      await subscription.on("invalid", () => {});
+      assert(false, "subscribing with invalid trigger");
+    } catch (e) {
+      await subscription.removeAllListeners();
+      assert(true);
+    }
+  });
+
+  it("should throw when subscribing without callback", async () => {
+    simpleStorage = simpleStorage.setWallet(wallet);
+    const subscription = simpleStorage.setValue("hello world");
+
+    try {
+      await subscription.on("confirmation");
+      assert(false, "subscribing without a callback");
+    } catch (e) {
+      await subscription.removeAllListeners();
+      assert(true);
+    }
+  });
+
+  it("should call a write contract method (send tx)", async () => {
```

Comment:
We should add a test case where that captures the case where the transaction never reaches 7 confirmations, and the user has subscribed to that event after a certain number of new blocks

Reply:
What should happen when never reach confirmations: Remove listener after X seconds?

Reply:
https://github.com/ethers-io/ethers.js/issues/385#issuecomment-450702497